core/src/main/java/de/danoeh/antennapod/core/service/playback/LocalPSMP.java

```diff
@@ -57,16 +58,42 @@
     private final ReentrantLock playerLock;
     private CountDownLatch seekLatch;
 
-    private final ThreadPoolExecutor executor;
+    private final PlayerExecutor executor;
+
+    /**
+     * All ExoPlayer methods must be executed on the same thread.
+     * We use the main application thread. This class allows to
+     * "fake" an executor that just calls the method instead of
+     * submitting to another thread. Other players are still
+     * executed in a background thread.
+     */
+    private class PlayerExecutor {
+        private boolean useMainThread = true;
+        private ThreadPoolExecutor threadPool;
+
+        public Future<?> submit(Runnable r) {
+            if (useMainThread) {
+                r.run();
```

Comment:
The logic `r.run()` here assumes the caller is calling from main thread. It may very well be indeed the case, but I'd be a tad weary.

Reply:
I agree. The problem I faced with posting handlers to the ui thread explicitly is that there are deadlocks. This way, it definitely is an improvement (all calls I found are indeed on ui thread). ExoPlayer might still complain in cases I have overlooked but that needs to be fixed on the caller side anyway.

Do you have another suggestion? Rename to useCallerThread? Add comment?